File: Packages/mcp-unity/Editor/Tools/ProfilerCaptureFrameTool.cs
using UnityEngine;
using UnityEditor;
using Newtonsoft.Json.Linq;
using McpUnity.Unity;
using McpUnity.Utils;

namespace McpUnity.Tools
{
    /// <summary>
    /// Tool to capture the current frame's performance data.
    /// Returns deltaTime, FPS, frame count, and time since startup.
    /// </summary>
    public class ProfilerCaptureFrameTool : McpToolBase
    {
        public ProfilerCaptureFrameTool()
        {
            Name = "profiler_capture_frame";
            Description = "Captures the current frame's timing info including deltaTime, FPS, and frame counts.";
        }

        public override JObject Execute(JObject parameters)
        {
            try
            {
                return new JObject
                {
                    ["success"] = true,
                    ["type"] = "text",
                    ["message"] = "Captured current frame data",
                    ["deltaTime"] = Time.deltaTime,
                    ["smoothDeltaTime"] = Time.smoothDeltaTime,
                    ["fps"] = System.Math.Round(1.0f / Time.smoothDeltaTime, 1),
                    ["frameCount"] = Time.frameCount,
                    ["timeSinceStartup"] = Time.realtimeSinceStartup,
                    ["timeScale"] = Time.timeScale,
                    ["isPlaying"] = Application.isPlaying
                };
            }
            catch (System.Exception ex)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    $"Error capturing frame data: {ex.Message}",
                    "profiler_error"
                );
            }
        }
    }
}

File: Packages/mcp-unity/Server~/src/tools/profilerCaptureFrameTool.ts
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnity } from '../unity/mcpUnity.js';
import { Logger } from '../utils/logger.js';
import { z } from 'zod';

export function registerProfilerCaptureFrameTool(
  server: McpServer,
  mcpUnity: McpUnity,
  logger: Logger
) {
  server.tool(
    'profiler_capture_frame',
    'Captures the current frame timing data including deltaTime, FPS, frame count, time since startup, and time scale.',
    {
      // No parameters needed
    },
    async (_params) => {
      logger.info('Capturing frame data');
      return await mcpUnity.sendRequest({
        method: 'profiler_capture_frame',
        params: {}
      });
    }
  );
}
